File: concurrency/src/org/osgi/book/reader/tutorial/JLabelMailboxCountTracker.java
package org.osgi.book.reader.tutorial;

import java.awt.EventQueue;

import javax.swing.JLabel;
import javax.swing.SwingUtilities;

import org.osgi.book.reader.api.Mailbox;
import org.osgi.framework.BundleContext;
import org.osgi.framework.ServiceReference;
import org.osgi.util.tracker.ServiceTracker;

public class JLabelMailboxCountTracker extends ServiceTracker {

   private final JLabel label;
   private int count = 0;

   public JLabelMailboxCountTracker(JLabel label,
         BundleContext context) {
      super(context, Mailbox.class.getName(), null);
      this.label = label;
   }

   @Override
   public Object addingService(ServiceReference reference) {
      int displayCount;
      synchronized (this) {
         count++;
         displayCount = count;
      }
      updateDisplay(displayCount);
      return null;
   }

   @Override
   public void removedService(ServiceReference reference,
         Object service) {
      int displayCount;
      synchronized (this) {
         count--;
         displayCount = count;
      }
      updateDisplay(displayCount);
   }

   private void updateDisplay(final int displayCount) {
      Runnable action = new Runnable() {
         public void run() {
            label.setText("There are " + displayCount + " mailboxes");
         }
      };
      if(EventQueue.isDispatchThread()) {
         action.run();
      } else {
         SwingUtilities.invokeLater(action);
      }
   }
}

File: concurrency/scala_src/JLabelMailboxCountTracker.scala
package org.osgi.book.reader.api {
  class Mailbox {
  }
  
  object ScalaSwingUtils {
    import javax.swing.SwingUtilities
    def inEDT(action: => Unit) = 
      if (EventQueue.isDispatchThread())
        action
      else
        SwingUtilities.invokeLater(new Runnable() {def run = action})
  }
}

package org.osgi.book.reader.tutorial.scala {
  import javax.swing._
  import org.osgi.framework._
  import org.osgi.util.tracker._
  import org.osgi.book.reader.api._
  import org.osgi.book.reader.api.ScalaSwingUtils._
  
  class JLabelMailboxCountTracker(ctx: BundleContext, label: JLabel)
      extends ServiceTracker(ctx, classOf[Mailbox].getName, null) {
    
    private var count = 0: int
    
    override
    def addingService(ref: ServiceReference): AnyRef = {
      var displayCount = 0
      synchronized { count += 1; displayCount = count }
      
      inEDT(label.setText("foo"))
      null
    }
    
    override
    def removedService(ref: ServiceReference, service: AnyRef) = {
      var displayCount = 0
      synchronized { count -= 1; displayCount = count }
      
      inEDT(label.setText("foo"))
    }
  }
}


